### ISSUE TITLE
Module namespace organization

### ISSUE THREAD

**Issue by cceyda** [kornia/kornia#804]
### Cleaner namespaces?

Right-now all(?) functional and class methods are available through the main module namespace. 
So when I write `kornia.` and hit tab I get all the things... this is good in some cases like for autocompletion, but end up being confusing in some (ie. later while reading code)

I like the higher submodule categorizations (augmentation,color,feature,filters etc) they make sense, but within each submodule there are inconsistencies.
Below are some things I have noticed:

- Some modules have an `F` submodule some have `functional` and some just have the functional things on the main level. 
 -> would be better to unite consistently. Imo `functional` is better and if wanted people can just import it as `F`.

- Some modules have higher modules as submodules(?) so something like `k.augmentation.augmentation.F.functional.adjust_brightness` ends up being valid and confusing (Not that you would write that but still 😕 ) 
 
- Some functions are not where you would expect them example: `motion_blur` is not in `kornia.filters.blur` but in `kornia.filters.motion.motion_blur`, In this case it is nice to have it in `kornia.filters.motion_blur` (or maybe `kornia.filters.functional.motion_blur`) (Docs have  `motion_blur`  under blur, maybe just version differences?)
    - Honestly `filters` and `feature` modules are a bit confusing. Example the `gaussian_blur2d ` in  kornia.feature.responses.gaussian_blur2d. The module name called responses vs detectors in docs.
	
- Some operations don't have functional equivalents or vice versa (maybe intentionally? maybe just forgotten?) 

Using v0.4.2 & looking at latest docs

## Motivation

This might be more of a personal preference thing, idk. 
It makes it harder to see/find what functionality is there, especially when not familiar with the library. 
There is a lot of functionality in Kornia which I love ❤️ but would love it even more if it is organized a bit more.

## Pitch

- Put all functional things under submodule.functional on the main level and on the subsubmodule level they can be higher. Example `kornia.functional.rgb_to_bgr` & `kornia.color.rgb_to_bgr`

## Alternatives

Open to suggestions~ Would like to hear what everyone else thinks!
Or explanations on the logic behind the current system.

**Comment by shijianjian** [comment]
Agreed.

In general, we shall stop using the lazy implementation of module imports like ```from .filters import *```, which would import the unnecessary dependents into the high level. I think it is the main problem that is affecting the user experience.

> * Put all functional things under submodule.functional on the main level and on the subsubmodule level they can be higher. Example `kornia.functional.rgb_to_bgr` & `kornia.color.rgb_to_bgr`

I think the whole kornia lib shall be at the functional level. We might have a ```kornia.nn``` to make all modulized functions to make the whole lib cleaner.

**Comment by edgarriba** [comment]
@cceyda thanks for pointing this out. The trend we have been more or less following is to move everything to functionals except for the augmentations which we might want to promote their usage as modules to help with the optimization of the parameters.

The reason for having everything exposed to the top level is purely for convenience at the time to import the whole package (we started small :)) What would change for example from `kornia.functional.rgb_to_bgr` to `kornia.rgb_to_bgr` ? I personally prefer the second.

The case for motion blur, I agree, feel free to send a PR fixing it. However, consider to raise a deprecation warning since at this point we have to make this kind of changes "slowly" giving beforehand feedback to the users. Regarding `gaussian_blur2d`, what version do you use ? I believe it has been existing to `kornia.filters` for a while.

We would love to work on cleaning the modules organization to make a more stable API.
Do you use slack ? feel free to join us so that we can continue the discussion in there - here is the [`LINK `](https://join.slack.com/t/kornia/shared_invite/zt-csobk21g-CnydWe5fmvkcktIeRFGCEQ)

**Comment by edgarriba** [comment]
@cceyda @shijianjian @anguelos @ducha-aiki @anguelos @pmeier @Borda I would resume this conversation and see how what conclusions we get from that. Can we write down some opinions for this ?

**Comment by cceyda** [comment]
I'll get the ball rolling on the discussions~

**kornia.nn** -> for classes that implement nn.Module (all of them)

- Make class names more consistent any class that takes a `p=probability` parameter for example classes in kornia.augmentation should start with `Random` right now some of them don't for example `GaussianBlur`. That way they are easier to distinct from classes in kornia.enhance.

**kornia.functional** -> for all functionals.

- I suggest we merge `kornia.augmentation` & `kornia.enhance`. They are basically doing the same thing. Except:
  - I'm not a fan of functions taking a `params: Dict`  like:  `kornia.augmentation.apply_adjust_brightness(input: torch.Tensor, params: Dict[str, torch.Tensor])`. This makes it a lot less transparent to use, hard to guess which function takes what param, can't use autocomplete etc. I wonder if there was a reason for this, am I missing something?
  - As such I suggest we replace kornia.augmentation's apply_* functions with their `kornia.enhance` equivalents. removing the *apply_* prefix

For convenience we can keep exposing everything through in the main `kornia.nn` & `kornia.functional` namespaces for convenience. But I also like the submodule naming.

- I like `kornia.color,kornia.contrib,kornia.feature,kornia.filter,kornia.geometry,kornia.morphology`
  - Maybe some things in filter, like blurring can be moved to `kornia.augmentation` ? except filter2D,filter3D (they feel like kernel ops)
  - I haven't used kornia.feature & kornia.geometry & kornia.morphology so can't comment. But it makes sense to me as is
  - I like kornia.losses as is. same for utils.

These submodules should have their respective `.functional` and `.nn`. For example: `kornia.losses.functional`&`kornia.losses.nn`

- Also I suggest in the docstrings we start adding min,max ranges for parameters that make sense to add it . Similar to torchvision `contrast (float or tuple of python:float (min, max))`

I know if we do this it a whole lot of deprecations 🙈

**Comment by edgarriba** [comment]
I agree for kornia.augmentation having Random* naming.
For the Param in apply, I believe @shijianjian was planning to deprécate the exposition of that in the top level.
As for merging augmentation and enhance, do not agree since the nature of both modules are different. Augmentation as you pointed, implements random sampling over others methos across the library, while enhance is specific to image illumination transforms.
Functional vs nn, in principle all kornia-core (no augmentation) has a design principle to be functional (with very few exceptions). @shijianjian proposed to have a top kornia.nn, but maybe it’s worth having .nn per module. I wouldn’t go for .functional.

Let’s target this for 0.7/0.8 once the core and augmentations is more mature.

one more comment from @pmeier is about avoiding to have more than 3 namespace levels. We need to very where this happens.

**Comment by cceyda** [comment]
> As for merging augmentation and enhance, do not agree since the nature of both modules are different. Augmentation as you pointed, implements random sampling over others methos across the library, while enhance is specific to image illumination transforms.

That is true for the nn.modular stuff, but the functionals(the apply_* functions) in `.augmentation` are not random & are just wrappers around enhance. For example; `augmentation.apply_adjust_saturation==enhance.adjust_saturation`. As it should be,since functionals shouldn't be random. 
Don't see why not image illumination stuff can't be augmentations?

I like Kornia's current implementation logic template:

```py
class Something(nn.Module):# or whatever base class
    def __init__(self,...) -> None:
        super().__init__(...)

    def __repr__(self) -> str:
       ...

    def generate_parameters(self,...): # If have randomness
       ...

    def apply_transform(self, input: torch.Tensor, ...):  # forward
        return F.something(input,...) #pass to functional
```

**Comment by pmeier** [comment]
IMO this is just a definition issue and one that urgently needs to be addressed before we go ahead with the restructuring. Both definitions look reasonable:

> Augmentation transforms comprise everything that involves randomly applying transformations or transformations with random parameters to the input

and

> Augmentation transforms comprise everything that in the input pipeline before the input is used for training

IMO we should have such a definition for every module to improve the UX for users. Plus, it would probably a good idea to add something like this the docs to give a better overview.

**Comment by edgarriba** [comment]
See `kornia.augmentations` as a small separated library on top of other kornia-core (we even had the discussion to create a separated repo for that). `augmentation.apply_adjust_saturation` and similar should be deprecated /cc @shijianjian I think already did some work on that. In this specific case we have been trying to promote something like this:
```python
class RandomSaturation(AugmentationBase2D):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

    def generate_parameters(self, shape: torch.Size):
       return dict(saturation_factor=torch.rand(shape))

    def apply_transform(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
       return kornia.enhance.adjust_saturation(input, params['saturation_factor'])
```
I agree the we need better documentation and more examples.
